### ISSUE TITLE
Can't sync from genesis using celestia-appd main

### ISSUE THREAD

**Issue by rootulp** [celestiaorg/celestia-app#3167]
## Context

I'm trying to do a manual sync from genesis test using a celestia-appd binary build from main. 

## Problem

I can't get past block height 1. I'm observing this log:

```
2:57AM ERR Error in validation err="wrong Block.Header.AppHash.  Expected 869EBEF734A657323408029077A7B0F835DFBB771582BBCE4084D29D5661DFF8, got F43C8B84340E6803AC86AD934B03E7145E155F7E8E2E7E19412D6ED8A44CBF14" module=blockchain
2:57AM ERR Stopping peer for error err="blockchainReactor validation error: wrong Block.Header.AppHash.  Expected 869EBEF734A657323408029077A7B0F835DFBB771582BBCE4084D29D5661DFF8, got F43C8B84340E6803AC86AD934B03E7145E155F7E8E2E7E19412D6ED8A44CBF14" module=p2p peer={"Data":{},"Logger":{}}
```

More info if helpful

```
root@rootul-mainnet-consensus:~# celestia-appd status
{"NodeInfo":{"protocol_version":{"p2p":"8","block":"11","app":"1"},"id":"70d79b49793fafff37ae4e311a9a34e4ef095c86","listen_addr":"tcp://0.0.0.0:26656","network":"celestia","version":"1.0.0-rc0-564-g82703272","channels":"40202122233038606100","moniker":"node-name","other":{"tx_index":"off","rpc_address":"tcp://127.0.0.1:26657"}},"SyncInfo":{"latest_block_hash":"6BE39EFD10BA412A9DB5288488303F5DD32CF386707A5BEF33617F4C43301872","latest_app_hash":"E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855","latest_block_height":"1","latest_block_time":"2023-10-31T14:00:00Z","earliest_block_hash":"6BE39EFD10BA412A9DB5288488303F5DD32CF386707A5BEF33617F4C43301872","earliest_app_hash":"E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855","earliest_block_height":"1","earliest_block_time":"2023-10-31T14:00:00Z","catching_up":true},"ValidatorInfo":{"Address":"03F7E426EC6352C8C698FE6015FF0F469CCEA44E","PubKey":{"type":"tendermint/PubKeyEd25519","value":"euC6cVEg8YLaio1O8Iy/jNZtprr1qQW9kpTQzHGA3uk="},"VotingPower":"0"}}
```

I used the script in https://github.com/celestiaorg/celestia-app/pull/3166

**Comment by evan-forbes** [comment]
per sync, I think best unit tests for crossing versions are the best start. However, during the final debugging phase of syncing from scract, maybe the script binary that the script listed here could be modified to dump the state and the problematic block. This way the bug could be easily replicated.

**Comment by cmwaters** [comment]
I've started making some modifications to support changes to the store an application has (so we have the correct stores and module for each version as these affect the app hash).

As part of this I have added print statements so we can get the app hashes of each of the modules so we can compare main and v1.x

For `main`, I am getting the following:

```
committing store upgrade with version 1 and hash E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855
committing store mint with version 1 and hash 0DD7FE0AFC24F6B271F85E98C76F3CE5F5C370EA84748187777B3183EA77ADBB
committing store feegrant with version 1 and hash E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855
committing store gov with version 1 and hash 60CD137C1962ECAC616389D68034833F2921509C2D285AA5F5153997CE968A74
committing store staking with version 1 and hash 5C07A3C39782EA44C5F4E69BF890063D1D9EA2BD0DB341390CB7EE56B85A4B49
committing store mem_capability with version 0 and hash
committing store acc with version 1 and hash 4291091D026D523F1776B32647812F09EBA52FB9C3A0A7031B7104F5AFF06655
committing store distribution with version 1 and hash 30074E31536DD979ABCD8F89901C77E7AB49A3AED85856D1BDA9651FA6703452
committing store params with version 1 and hash 3C2D7003171D6B25790A84B5567C65754A88D651DC1BE19A8B93BC0A1249A914
committing store authz with version 1 and hash E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855
committing store transfer with version 1 and hash 259ACE33F3D819FD9CC751968CC83D503A92E8A7E6FD5E3298068249565FD706
committing store qgb with version 1 and hash 885A9DF9F6D753292C2B686E1CEAFB78786F3B7413FCAFE4ED68C0A1DFF25DAF
committing store evidence with version 1 and hash E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855
committing store capability with version 1 and hash 379646E1B21B0D39607965B0AD2371F83EECE289EB469D024DD8FFC3CFBB0CD0
committing store ibc with version 1 and hash C617EEF82D350859FC4C1B0118079FAD20ECD1E5FE2294C86A86E94C0D14CA55
committing store slashing with version 1 and hash 94FF25BC98BF33B6981D747A20A20F71A133D04071F64E8C6ED48108BD94C880
```

**Comment by cmwaters** [comment]
Ok latest update. Stores are the same. We have an apphash mismatch with upgrade:

`main`: `E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855`
`v1.x`: `EF4C6B20D316A771F8841921A97CEA58E911391A7A90E374E3A2BDB57DCC4522`

Note that for main, the hash is for an empty store

**Comment by cmwaters** [comment]
This turned out to be the problem:

https://github.com/celestiaorg/celestia-app/blob/91b6827a80ef8a4d0048ca81f9eda4c88f37a2ef/app/app.go#L577

Which btw is totally useless since we don't even use the upgrade keeper for that

**Comment by rootulp** [comment]
I started a Scaleway instance and tried syncing from genesis with `main`, specifically: a8e3085c98301d0dc0c2a5c86880b5fc51f904de. It fails on block 44. 

```
root@rootulp-consensus-main:~/celestia-app# celestia-appd status
{"NodeInfo":{"protocol_version":{"p2p":"8","block":"11","app":"1"},"id":"95ef83649f0c913d55751b1682754d07b253ab0f","listen_addr":"tcp://0.0.0.0:26656","network":"celestia","version":"1.0.0-rc0-648-ga8e3085c","channels":"40202122233038606100","moniker":"node-name","other":{"tx_index":"off","rpc_address":"tcp://127.0.0.1:26657"}},"SyncInfo":{"latest_block_hash":"8FB417D53BCBA7B99590665754039954D065EE2B757842FEE5C6A82388840C9B","latest_app_hash":"0CA5614B3810F3A9E08C1211A6B8E6B8FD8EA4B14B91B90F0E94A77FC5501554","latest_block_height":"44","latest_block_time":"2023-10-31T14:10:23.458314922Z","earliest_block_hash":"6BE39EFD10BA412A9DB5288488303F5DD32CF386707A5BEF33617F4C43301872","earliest_app_hash":"E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855","earliest_block_height":"1","earliest_block_time":"2023-10-31T14:00:00Z","catching_up":true},"ValidatorInfo":{"Address":"A201F0B761AC85390142F43C83A4007AAE7B6B29","PubKey":{"type":"tendermint/PubKeyEd25519","value":"gUwYN9FQxGziHD8KJ5RlslYnHd1puTWznQbyO2yPJQY="},"VotingPower":"0"}}
```

IDK why the logs look so weird but looks like an app hash mismatch:

```logs
^[[90m6:27PM^[[0m ^[[31mERR^[[0m ^[[1mError in validation^[[0m ^[[36merr=^[[0m"wrong Block.Header.AppHash.  Expected CFBFC8134A9D2ECCA6935737755B6FC71E8AF6E16E239CF00850EA82FB1733FB, got D4A358BA705AFD122AD4154FC17383878BDB810FCF41D8AD643020642D2555E9" ^[[36mmodule=^[[0mblockchain
^[[90m6:27PM^[[0m ^[[31mERR^[[0m ^[[1mStopping peer for error^[[0m ^[[36merr=^[[0m"blockchainReactor validation error: wrong Block.Header.AppHash.  Expected CFBFC8134A9D2ECCA6935737755B6FC71E8AF6E16E239CF00850EA82FB1733FB, got D4A358BA705AFD122AD4154FC17383878BDB810FCF41D8AD643020642D2555E9" ^[[36mmodule=^[[0mp2p ^[[36mpeer=^[[0m{"Data":{},"Logger":{}}
```

Scaleway instance: rootulp-consensus-main if you want to have a look.

**Comment by cmwaters** [comment]
Also got this error when statesyncing:

```
11:33AM INF Applied snapshot chunk to ABCI app chunk=54 format=2 height=1258500 module=statesync total=56
11:33AM ERR failed to restore snapshot err="multistore restore: cannot import into non-IAVL store \"acc\": internal logic error"
11:33AM INF Applied snapshot chunk to ABCI app chunk=55 format=2 height=1258500 module=statesync total=56
11:33AM ERR State sync failed err="state sync aborted" module=statesync
```

**Comment by cmwaters** [comment]
Can confirm that I got the exact same app hash mismatch at height 44. I will start posting my progress on it here.

**Comment by cmwaters** [comment]
Ok so module hashes are as follows:

For `main`:

```
key KVStoreKey{0xc0010ba1e0, acc} hash D15BFF423553FBC6B2F5DF05087DC756EC5256B5BF7DA3F3D92C7BB9AF1FE295
key KVStoreKey{0xc0010ba1f0, authz} hash E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855
key KVStoreKey{0xc0010ba200, bank} hash 672309A593A157B6371E46C448EA955ECC51849968813D50D29FBAE1D8BF9FD4
key KVStoreKey{0xc0010ba210, staking} hash 49E515A45D96E3FFA4B3031FD5F0F6212127CDEF0592158F326D8BE6A489A84A
key KVStoreKey{0xc0010ba220, mint} hash DC04EEEBDC5ECBECB9938E3439E978BA3B9EB26B55F96C5880D2589FE8F51C56
key KVStoreKey{0xc0010ba230, distribution} hash F76EFF94C9F3B59223D9CE1CDE5E79F8FB2065B81375FB396E0A01DE2FBAA53F
key KVStoreKey{0xc0010ba240, slashing} hash 7F30F93BFAF67B91E589792FD3D31D35842951D36DFF203E91B104A57462BDB7
key KVStoreKey{0xc0010ba250, gov} hash 60CD137C1962ECAC616389D68034833F2921509C2D285AA5F5153997CE968A74
key KVStoreKey{0xc0010ba260, params} hash 3C2D7003171D6B25790A84B5567C65754A88D651DC1BE19A8B93BC0A1249A914
key KVStoreKey{0xc0010ba270, upgrade} hash EF4C6B20D316A771F8841921A97CEA58E911391A7A90E374E3A2BDB57DCC4522
key KVStoreKey{0xc0010ba280, feegrant} hash E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855
key KVStoreKey{0xc0010ba290, evidence} hash E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855
key KVStoreKey{0xc0010ba2a0, capability} hash 387F2FDB4F615D6C5D919F4D951CE12B308C1E0D1CE254EFD110DED24DC0FD5D
key KVStoreKey{0xc0010ba310, blob} hash E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855
key KVStoreKey{0xc0010ba2b0, qgb} hash 5CFE85C0A142023C585824D63D91A6BEEEFD778167A4DD9C739B3C26715B9D4B
key KVStoreKey{0xc0010ba2c0, transfer} hash 259ACE33F3D819FD9CC751968CC83D503A92E8A7E6FD5E3298068249565FD706
key KVStoreKey{0xc0010ba2d0, ibc} hash 326E9EC2DB91E94752DF756D8DC12553DA444AF0BFB8CB3C6031691C3F17BFB1
key MemoryStoreKey{0xc0010ba350, mem_capability} hash
```

For `v1.8.0`:

```
key KVStoreKey{0xc0016ec600, acc} hash D15BFF423553FBC6B2F5DF05087DC756EC5256B5BF7DA3F3D92C7BB9AF1FE295
key KVStoreKey{0xc0016ec610, authz} hash E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855
key KVStoreKey{0xc0016ec620, bank} hash 672309A593A157B6371E46C448EA955ECC51849968813D50D29FBAE1D8BF9FD4
key KVStoreKey{0xc0016ec630, staking} hash 49E515A45D96E3FFA4B3031FD5F0F6212127CDEF0592158F326D8BE6A489A84A
key KVStoreKey{0xc0016ec640, mint} hash DC04EEEBDC5ECBECB9938E3439E978BA3B9EB26B55F96C5880D2589FE8F51C56
key KVStoreKey{0xc0016ec650, distribution} hash F76EFF94C9F3B59223D9CE1CDE5E79F8FB2065B81375FB396E0A01DE2FBAA53F
key KVStoreKey{0xc0016ec660, slashing} hash 7F30F93BFAF67B91E589792FD3D31D35842951D36DFF203E91B104A57462BDB7
key KVStoreKey{0xc0016ec670, gov} hash 60CD137C1962ECAC616389D68034833F2921509C2D285AA5F5153997CE968A74
key KVStoreKey{0xc0016ec680, params} hash 3C2D7003171D6B25790A84B5567C65754A88D651DC1BE19A8B93BC0A1249A914
key KVStoreKey{0xc0016ec690, upgrade} hash EF4C6B20D316A771F8841921A97CEA58E911391A7A90E374E3A2BDB57DCC4522
key KVStoreKey{0xc0016ec6a0, feegrant} hash E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855
key KVStoreKey{0xc0016ec6b0, evidence} hash E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855
key KVStoreKey{0xc0016ec6c0, capability} hash 387F2FDB4F615D6C5D919F4D951CE12B308C1E0D1CE254EFD110DED24DC0FD5D
key KVStoreKey{0xc0016ec6d0, blob} hash E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855
key KVStoreKey{0xc0016ec6e0, qgb} hash 5CFE85C0A142023C585824D63D91A6BEEEFD778167A4DD9C739B3C26715B9D4B
key KVStoreKey{0xc0016ec6f0, transfer} hash 259ACE33F3D819FD9CC751968CC83D503A92E8A7E6FD5E3298068249565FD706
key KVStoreKey{0xc0016ec700, ibc} hash 41499EF14A0DD77B8FD224BD2E6ECFABBCD3FC44337F61177C91D5EF61530690
key MemoryStoreKey{0xc0016ec740, mem_capability} hash
```

The culprit is `ibc`:

For `main`:

It is `326E9EC2DB91E94752DF756D8DC12553DA444AF0BFB8CB3C6031691C3F17BFB1`

For `v1.8.0`:

It is `41499EF14A0DD77B8FD224BD2E6ECFABBCD3FC44337F61177C91D5EF61530690`

The main culprit is probably PFM which runs for both v1 and v2 (since it's middleware and not actually versioned)

**Comment by cmwaters** [comment]
https://github.com/celestiaorg/celestia-app/pull/3353 seems to resolve the current app hash mismatch. @rootulp are you able to try rerun your node using this commit over the weekend to see if we bump into any more problems

**Comment by rootulp** [comment]
Just an idea: we expect syncing from genesis to the current network tip to be slow so we can collect more info on this by running these in parallel:
1. A node syncing from genesis to current network tip
2. A node syncing via some recent snapshot to current network tip

**Comment by rootulp** [comment]
Node stopped some time over night. When I try to start it:

```
root@rootulp-consensus-main:~# celestia-appd start
2:37PM INF starting node with ABCI Tendermint in-process
2:37PM INF service start impl=multiAppConn module=proxy msg={}
2:37PM INF service start connection=query impl=localClient module=abci-client msg={}
2:37PM INF service start connection=snapshot impl=localClient module=abci-client msg={}
2:37PM INF service start connection=mempool impl=localClient module=abci-client msg={}
2:37PM INF service start connection=consensus impl=localClient module=abci-client msg={}
2:37PM INF service start impl=EventBus module=events msg={}
2:37PM INF service start impl=PubSub module=pubsub msg={}
2:37PM INF service start impl=IndexerService module=txindex msg={}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xf11e96]

goroutine 7 [running]:
github.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).createQueryContext(_, _, _)
	/root/go/pkg/mod/github.com/celestiaorg/cosmos-sdk@v1.21.0-sdk-v0.46.16/baseapp/abci.go:716 +0x1f6
github.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).Info(0xc0000fcd20, {{0x78?, 0x18ed9c5?}, 0x0?, 0x0?})
	/root/go/pkg/mod/github.com/celestiaorg/cosmos-sdk@v1.21.0-sdk-v0.46.16/baseapp/abci.go:134 +0xe5
github.com/celestiaorg/celestia-app/v2/app.(*App).Info(0xc00106ca88, {{0x2595993?, 0x10?}, 0xc000606008?, 0xc000645e40?})
	/root/celestia-app/app/app.go:500 +0x56
github.com/tendermint/tendermint/abci/client.(*localClient).InfoSync(0xc0013dac60, {{0x2595993?, 0x1?}, 0xc0007438d0?, 0x41565b?})
	/root/go/pkg/mod/github.com/celestiaorg/celestia-core@v1.35.0-tm-v0.34.29/abci/client/local_client.go:250 +0x109
github.com/tendermint/tendermint/proxy.(*appConnQuery).InfoSync(0xc00107a400?, {{0x2595993?, 0xc000645e30?}, 0x213ffc0?, 0x7f73ee35fb98?})
	/root/go/pkg/mod/github.com/celestiaorg/celestia-core@v1.35.0-tm-v0.34.29/proxy/app_conn.go:170 +0x23
github.com/tendermint/tendermint/consensus.(*Handshaker).HandshakeWithContext(0xc000743cd8, {0x321fe38, 0x46f7540}, {0x3236fd0, 0xc0005300d0})
	/root/go/pkg/mod/github.com/celestiaorg/celestia-core@v1.35.0-tm-v0.34.29/consensus/replay.go:250 +0x7a
github.com/tendermint/tendermint/consensus.(*Handshaker).Handshake(...)
	/root/go/pkg/mod/github.com/celestiaorg/celestia-core@v1.35.0-tm-v0.34.29/consensus/replay.go:243
github.com/tendermint/tendermint/node.doHandshake({_, _}, {_, _}, {{{0xb, 0x1}, {0xc0011a23c0, 0x7}}, {0xc0011a23c8, 0x8}, ...}, ...)
	/root/go/pkg/mod/github.com/celestiaorg/celestia-core@v1.35.0-tm-v0.34.29/node/node.go:337 +0x1aa
github.com/tendermint/tendermint/node.NewNode(0xc001434140, {0x32146a0, 0xc00078fae0}, 0xc00105ff30, {0x31fdea0, 0xc00119b8d8}, 0xc000745980, 0x2c707b0, 0xc000745970, {0x321fc40, ...}, ...)
	/root/go/pkg/mod/github.com/celestiaorg/celestia-core@v1.35.0-tm-v0.34.29/node/node.go:831 +0x6ab
github.com/cosmos/cosmos-sdk/server.startInProcess(_, {{0x0, 0x0, 0x0}, {0x3247590, 0xc0013e8a50}, 0x0, {0xc0013ea8e8, 0x8}, {0x323ad00, ...}, ...}, ...)
	/root/go/pkg/mod/github.com/celestiaorg/cosmos-sdk@v1.21.0-sdk-v0.46.16/server/start.go:301 +0x745
github.com/cosmos/cosmos-sdk/server.StartCmd.func2.2()
	/root/go/pkg/mod/github.com/celestiaorg/cosmos-sdk@v1.21.0-sdk-v0.46.16/server/start.go:147 +0x45
github.com/cosmos/cosmos-sdk/server.wrapCPUProfile.func2()
	/root/go/pkg/mod/github.com/celestiaorg/cosmos-sdk@v1.21.0-sdk-v0.46.16/server/start.go:535 +0x23
created by github.com/cosmos/cosmos-sdk/server.wrapCPUProfile in goroutine 1
	/root/go/pkg/mod/github.com/celestiaorg/cosmos-sdk@v1.21.0-sdk-v0.46.16/server/start.go:534 +0x23f
```

which is the same error as before. Coming from [this line](https://github.com/celestiaorg/celestia-app/blob/32accbaa27cf89cd03da382d551d76888c7bddc1/app/app.go#L500). I'm restarting after a `tendermint unsafe-reset-all` but I think there's a lingering issue here.

**Comment by cmwaters** [comment]
I have a fix for this issue. Should be ready tomorrow

**Comment by rootulp** [comment]
I got to block height 685 with the versioned IBC middleware fix but couldn't restart the node after a `panic`. Attempting to sync a new node from genesis based on the current `main` which contains IBC middleware fix + restart node fix.

**Comment by rootulp** [comment]
Confirmed there is an issue at block height 685

```
2:38AM ERR Error in validation err="wrong Block.Header.LastResultsHash.  Expected A9360C6F83F4D5869D3C51D774467C4735F901EB4C5FD1E112F53D8F76734A81, got 0442CACE9C7FE831F06DECEBF1A26DF7ADFC4AA014F36DCF7CD65D67297D28F7" module=blockchain
2:38AM ERR Stopping peer for error err="blockchainReactor validation error: wrong Block.Header.LastResultsHash.  Expected A9360C6F83F4D5869D3C51D774467C4735F901EB4C5FD1E112F53D8F76734A81, got 0442CACE9C7FE831F06DECEBF1A26DF7ADFC4AA014F36DCF7CD65D67297D28F7" module=p2p peer={"Data":{},"Logger":{}}
```

**Comment by rootulp** [comment]
`LastResultsHash` is based on these significant fields:

```
type ResponseDeliverTx struct {
	Code      uint32  `protobuf:"varint,1,opt,name=code,proto3" json:"code,omitempty"`
	Data      []byte  `protobuf:"bytes,2,opt,name=data,proto3" json:"data,omitempty"`
	GasWanted int64   `protobuf:"varint,5,opt,name=gas_wanted,proto3" json:"gas_wanted,omitempty"`
	GasUsed   int64   `protobuf:"varint,6,opt,name=gas_used,proto3" json:"gas_used,omitempty"`
}
```

For block 684, there were 97 transactions so 97 `ResponseDeliverTx`s to analyze and compare to the output from a `main` node. 

Ref: https://www.mintscan.io/celestia/block/684?sector=events

One way to debug is to write a Go tool that will read the state/store/db, call LoadABCIResponses(684) => json.Marshal => print out. Compare to v1.x.

**Comment by rootulp** [comment]
I thought of an alternative to the above and that is to just query the Tendermint RPC for my node via 

```shell
curl -X GET http://127.0.0.1:26657/block_results?height=684
```

Unfortunately, I didn't have ABCI responses persisted so I had to do:

1. Set `discard_abci_responses: false`
2. Rollback to the previous block height:
```
root@rootulp-consensus-main:~# celestia-appd rollback
Rolled back state to height 684 and hash 6340DFC7EA7BA4E43A13B74646006E343AE52444B52519A4E3E5FCAFB09EAFC3
```
3. Start the node up again. Unfortunately that doesn't work because:

```
panic: loading latest version: failed to load latest version: version of store ibc mismatch root store's version; expected 684 got 0

goroutine 12 [running]:
github.com/celestiaorg/celestia-app/v2/app.(*App).Info(0xc00143f508, {{0x2588830?, 0x10?}, 0xc000100008?, 0xc001525450?})
	/root/celestia-app/app/app.go:519 +0x49c
github.com/tendermint/tendermint/abci/client.(*localClient).InfoSync(0xc0014802a0, {{0x2588830?, 0x1?}, 0xc0010d58d0?, 0x41565b?})
	/root/go/pkg/mod/github.com/celestiaorg/celestia-core@v1.35.0-tm-v0.34.29/abci/client/local_client.go:250 +0x109
github.com/tendermint/tendermint/proxy.(*appConnQuery).InfoSync(0xc0012f8600?, {{0x2588830?, 0xc001525440?}, 0x2134040?, 0x7f4ac45d7338?})
	/root/go/pkg/mod/github.com/celestiaorg/celestia-core@v1.35.0-tm-v0.34.29/proxy/app_conn.go:170 +0x23
github.com/tendermint/tendermint/consensus.(*Handshaker).HandshakeWithContext(0xc0010d5cd8, {0x3212268, 0x46e23e0}, {0x3228d50, 0xc00190cf70})
	/root/go/pkg/mod/github.com/celestiaorg/celestia-core@v1.35.0-tm-v0.34.29/consensus/replay.go:250 +0x7a
github.com/tendermint/tendermint/consensus.(*Handshaker).Handshake(...)
	/root/go/pkg/mod/github.com/celestiaorg/celestia-core@v1.35.0-tm-v0.34.29/consensus/replay.go:243
github.com/tendermint/tendermint/node.doHandshake({_, _}, {_, _}, {{{0xb, 0x1}, {0xc00141cfb0, 0x7}}, {0xc00141cfb8, 0x8}, ...}, ...)
	/root/go/pkg/mod/github.com/celestiaorg/celestia-core@v1.35.0-tm-v0.34.29/node/node.go:337 +0x1aa
github.com/tendermint/tendermint/node.NewNode(0xc001464140, {0x3206b00, 0xc0015b80a0}, 0xc001525ae0, {0x31f0300, 0xc0015ae6c0}, 0xc0010d7980, 0x2c635e8, 0xc0010d7970, {0x3212070, ...}, ...)
	/root/go/pkg/mod/github.com/celestiaorg/celestia-core@v1.35.0-tm-v0.34.29/node/node.go:831 +0x6ab
github.com/cosmos/cosmos-sdk/server.startInProcess(_, {{0x0, 0x0, 0x0}, {0x3239300, 0xc0014188d0}, 0x0, {0xc00141c858, 0x8}, {0x322ca80, ...}, ...}, ...)
	/root/go/pkg/mod/github.com/celestiaorg/cosmos-sdk@v1.22.0-sdk-v0.46.16/server/start.go:301 +0x745
github.com/cosmos/cosmos-sdk/server.StartCmd.func2.2()
	/root/go/pkg/mod/github.com/celestiaorg/cosmos-sdk@v1.22.0-sdk-v0.46.16/server/start.go:147 +0x45
github.com/cosmos/cosmos-sdk/server.wrapCPUProfile.func2()
	/root/go/pkg/mod/github.com/celestiaorg/cosmos-sdk@v1.22.0-sdk-v0.46.16/server/start.go:535 +0x23
created by github.com/cosmos/cosmos-sdk/server.wrapCPUProfile in goroutine 1
	/root/go/pkg/mod/github.com/celestiaorg/cosmos-sdk@v1.22.0-sdk-v0.46.16/server/start.go:534 +0x23f
```

which seems like a separate issue worth investigating 😢

**Comment by rootulp** [comment]
Turns out the diff was not in 684 but it is in 685.

Ref:
- 684 diff is identical: https://www.diffchecker.com/QpV0nhDL/
- 685 diff has changes: https://www.diffchecker.com/evIJxGc8/ left is v1, right is v2

This one looks interesting:

![Screenshot 2024-04-30 at 7 47 26 PM](https://github.com/celestiaorg/celestia-app/assets/3699047/9ac646fb-6da0-44d5-9b00-195b54c717d5)

**Comment by rootulp** [comment]
The problematic transaction is [here](https://www.mintscan.io/celestia/tx/2CF9B3A8E0475416EACFA770B1CF438DB6D3C657BE048C33374A132AF5174BC6?height=685&sector=logs) which fails due to out of gas for an IBC timeout.

celestia-app version | ibc-go version
---------------------|---------------
1.0.0                | v6.2.0
1.8.0                | v6.3.0
main                 | v6.3.1

It looks like there may have been a breaking change to the gas schedule between [v6.2.0 and v6.3.0](https://github.com/cosmos/ibc-go/compare/v6.2.0...v6.3.0). I'm going to try to reproduce the error by syncing from genesis via v1.8.0.

**Comment by rootulp** [comment]
Trying to debug block height 687 now, module stores:

```
6:57PM INF executed block height=687 module=consensus num_invalid_txs=9 num_valid_txs=105
key KVStoreKey{0x140006bb930, upgrade}, commitID.Hash EF4C6B20D316A771F8841921A97CEA58E911391A7A90E374E3A2BDB57DCC4522
key KVStoreKey{0x140006bb6f0, acc}, commitID.Hash 26266B87B404B7FFE8FD2219FCE4A1A859712E94D1883D097026B688B8B95A9C
key KVStoreKey{0x140006bb920, params}, commitID.Hash 3C2D7003171D6B25790A84B5567C65754A88D651DC1BE19A8B93BC0A1249A914
key MemoryStoreKey{0x140006bb9e0, mem_capability}, commitID.Hash
key KVStoreKey{0x140006bb960, capability}, commitID.Hash FAC6E6578AC79F5DB047037A56225AE9E65477E074A5039D37D4A946E62F919A
key KVStoreKey{0x140006bb900, slashing}, commitID.Hash C2F4977FB974D3D8F5140F095E4C3227F0BF4037E5EE308D60142CC22324C05D
key KVStoreKey{0x140006bb940, feegrant}, commitID.Hash E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855
key KVStoreKey{0x140006bb990, transfer}, commitID.Hash 259ACE33F3D819FD9CC751968CC83D503A92E8A7E6FD5E3298068249565FD706
key KVStoreKey{0x140006bb970, blob}, commitID.Hash E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855
key KVStoreKey{0x140006bb9a0, ibc}, commitID.Hash C9DD26D3792C68112996D2F57F41E9DAC9ADD7B21E5D26AE3A350B6A83EB8774
key KVStoreKey{0x140006bb8f0, distribution}, commitID.Hash 6E59E5E88DB224AFD447871B0159AE66A145A6989F36A1208B0CE41223B38684
key KVStoreKey{0x140006bb910, gov}, commitID.Hash 60CD137C1962ECAC616389D68034833F2921509C2D285AA5F5153997CE968A74
key KVStoreKey{0x140006bb720, staking}, commitID.Hash 08DB861D5F93ACC24C695E742D795793BEC7031FBE5F56860032AD9CED4EA71A
key KVStoreKey{0x140006bb700, authz}, commitID.Hash 2EFC433F495987A0873362A6514DC1ECC421AD8410E271A96969C1EF8B42E9C4
key KVStoreKey{0x140006bb840, mint}, commitID.Hash 71B08A580D3287A4F9E772DB1D228B10091BDE213118FCFC7198FE3125F9E214
key KVStoreKey{0x140006bb710, bank}, commitID.Hash 5E36A8AA46CB9DD7A42314A726D746B83128477AF804DEE4ACCC3B97144F7CCB
key KVStoreKey{0x140006bb950, evidence}, commitID.Hash E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855
key KVStoreKey{0x140006bb980, qgb}, commitID.Hash 2CA9324319C39101FECE0C1C006FF62BC3D0EAC67B7B62988880CFFB31F89691
```

**Comment by rootulp** [comment]
It looks like the app hash mismatch on block height 687 is only reproducible using my blockstore.db + state.db because Callum and I could repro it with them but no one else could repro it. I'm running two nodes overnight:
1. rootulp-consensus: uses v1.x after a `tendermint unsafe-reset-all`. If this works, we'll cut v1.9.0
2. rootulp-consensus-main: uses v1.7.0 after a `tendermint unsafe-reset-all`. If this works, we don't need to redact v1.7.0.

**Comment by rootulp** [comment]
Update: 
- 1 worked so we cut celestia-app v1.9.0.
- 2 worked (currently syncing and at height 474095) so we don't need to redact v1.7.0.
- Also synced a node from `main` from genesis and it's at height 535,541 and still making progress.

**Comment by rootulp** [comment]
Update:
- v1.9.0 node is at height 610,058 and still making progress.
- `main` node is at height 689,343 and still making progress.

**Comment by rootulp** [comment]
Update:

- v1.9.0 node is at height 637,746 and still making progress.
- `main` node is at height 717,497 but hasn't made progress in a while. It seems stuck.

Logs

```
11:18PM INF executed block height=717497 module=state num_invalid_txs=0 num_valid_txs=13
11:18PM INF updates to validators module=state updates=05D2E22C1216CD30104BCDB6B122B1E7F1ACB5CD:25031101,64A69907D6AE18450250C7820F7C6776C2C06FC6:1412700,54FD50A55FEDC0D9338421AB32F08C9A9AD66CBE:1053372,5AFFF88E3E256A84A0C9EB348B044CADB912197B:870522
11:18PM INF commit synced commit=436F6D6D697449447B5B383020323338203237203237203137203139342034312032333920363620313338203134392031303220313136203134203132342031313620313932203134352032313620323535203832203333203232352032303220373420383420313934203431203438203139203139382039325D3A41463242397D
11:18PM INF committed state app_hash=50EE1B1B11C229EF428A9566740E7C74C091D8FF5221E1CA4A54C2293013C65C height=717497 module=state num_txs=13
11:18PM INF indexed block exents height=717497 module=txindex
```

and then repeatedly prints out:

```
11:18PM INF Ensure peers module=pex numDialing=0 numInPeers=0 numOutPeers=10 numToDial=0
11:19PM INF Saving AddrBook to file book=/root/.celestia-app/config/addrbook.json module=p2p size=1011
```

**Comment by rootulp** [comment]
Two options to resolve the issue where `main` is stuck:
1. Restart the node
1. Wipe the address book

**Comment by rootulp** [comment]
Update:
- v1.9.0 node is at height 717,054 and making progress
- main node is at height 779,613 and making progress

**Comment by evan-forbes** [comment]
the block sync is so :sloth:, and not in a good way lol

**Comment by rootulp** [comment]
I did take a Scaleway image of the main node around block height 700,000 and I'll repeat once we're at the tip. 

That way, if we want to test syncing from scratch again we can spin up a node using the image and need to rely on block sync. I think we can `rm -rf application.db` and start the node to execute all the blocks that it has locally.

**Comment by rootulp** [comment]
Update:
- v1.9.0 node is at height 744,547 and making progress
- main node is at height 918,488 and making progress

**Comment by rootulp** [comment]
Update:
- v1.9.0 node was at stuck at height 766,534 and not making progress. I restarted it and now it's continuing. 
- main node is at height 1,213,853 and making progress

**Comment by rootulp** [comment]
Update:
- v1.9.0 node was at height 1242121 and not making progress so I restarted it
- main node was at height 1,216,547 and not making progress so I restarted it

**Comment by rootulp** [comment]
Update:
- v1.9.0 node is at height 1,292,149 and making progress
- main node is at height 1,268,427 and making progress

**Comment by rootulp** [comment]
Update:

- v1.9.0 node ran out of disk space. `panic: failed to write batch: write /root/.celestia-app/data/application.db/46474675.log: no space left on device`. It was provisioned with a 1TB block storage but only 900GB seems usable on the instance. I just increased it to 1.2 TB. Ran `halt` then rebooted instance. Then:

```
rm mainnet.log
celestia-appd experimental-compact-goleveldb
lsblk
sudo growpart /dev/sda 1
sudo resize2fs /dev/sda1
```

then started node again. Same problem on main node so I increased it to 1.3 TB.


- v1.9.0 node is at height 1,462,582 and making progress.
- main node is at height 1,449,449 and making progress.

**Comment by rootulp** [comment]
Update:
- v1.9.0 node is at height 1,700,429 and making progress.
- main node is at height 1,700,425 and making progress.

**Comment by rootulp** [comment]
I'm going to:
- [ ] stop celestia-appd on the main node
- [ ] take a Scaleway snapshot of the main node
- [ ] wipe the application.db on the main node
- [ ] fetch the latest `main` because the last commit on the main node is 0f1bf6d6a292a3cbaccfd0c7ef3299531a0c5187
- [ ] start it again to see if it can sync from genesis much faster with a local state.db and blockstore.db that's already populated
